core/utils/debug.py

```diff
@@ -0,0 +1,106 @@
+            return self.root_element.find_elements(by_expression, value)
+
+    def get_shadow_element_in_shadow_dom(self, value, shadow_dom_root_element=None):
+        if shadow_dom_root_element:
+            return self.driver.execute_script(self.shadow_inner_element_script, shadow_dom_root_element, value)
+        else:
+            return self.driver.execute_script(self.shadow_inner_element_script, self.root_element, value)
+
+    def get_element_by_css_selector_and_text(self, value, text, shadow_dom_root_element=None, contains_text=False):
+        if shadow_dom_root_element:
+            for element in shadow_dom_root_element.find_elements(By.CSS_SELECTOR, value):
+                if contains_text:
+                    if text in (''.join(element.text)).encode('utf-8'):
+                        return element
+                else:
+                    if (''.join(element.text)).encode('utf-8') == text:
+                        return element
+        else:
+            for element in self.root_element.find_elements(By.CSS_SELECTOR, value):
+                if contains_text:
+                    if text in (''.join(element.text)).encode('utf-8'):
+                        return element
+                else:
+                    if (''.join(element.text)).encode('utf-8') == text:
+                        return element
+        return None
+
+    def wait_until_shadow_dom_element_located(self, by_expression, value, shadow_dom_root_element=None,
+                                              assert_element=True, timeout=60):
+        test_result = Wait.until(
+            lambda: len(self.get_elements_in_shadow_dom(by_expression, value, shadow_dom_root_element)) != 0,
+            timeout=timeout,
+            period=5)
+        if assert_element:
+            assert test_result, "Element searched by " + str(
+                by_expression) + " with value " + value + " not found for " + str(timeout) + " !"
+        elif test_result:
+            Log.info(
+                "Element " + value + " by expression " + by_expression + " for " + str(timeout) + " is not located!")
+
+    def reconnect_to_devtools(self):
+        self.wait_until_element_is_visible(By.CSS_SELECTOR, 'div.vbox.flex-auto.dimmed-pane', timeout=15)
+        shadow_dom_element = self.driver.find_elements(By.CSS_SELECTOR, 'div.vbox.flex-auto.dimmed-pane')
+        if shadow_dom_element:
+            shadow_div_element = self.driver.execute_script(self.shadow_root_element_script, shadow_dom_element[0])
+            shadow_elements = self.get_shadow_element_in_shadow_dom(".vbox.flex-auto", shadow_div_element)
+            reconnect_button = self.get_element_by_css_selector_and_text("button", "Reconnect DevTools",
+                                                                         shadow_elements)
+            if reconnect_button is not None:
+                reconnect_button.click()
+
+    def wait_until_element_is_visible(self, by_expression, value, timeout=15):
```

Comment:
Can we move it to chrome since it is not debugger specific.

Reply:
done